source/extensions/filters/network/ratelimit/ratelimit.h

```diff
@@ -83,13 +84,42 @@ class Filter : public Network::ReadFilter,
   void initializeReadFilterCallbacks(Network::ReadFilterCallbacks& callbacks) override {
     filter_callbacks_ = &callbacks;
     filter_callbacks_->connection().addConnectionCallbacks(*this);
+    injectDynamicDescriptorKeys(config_->descriptors(), filter_callbacks_->connection());
+  }
+
+  void injectDynamicDescriptorKeys(std::vector<RateLimit::Descriptor> original_descriptors,
+                                   const Network::Connection& connection) {
+    std::vector<RateLimit::Descriptor> dynamicDescriptors = std::vector<RateLimit::Descriptor>();
+    for (const RateLimit::Descriptor& descriptor : original_descriptors) {
+      RateLimit::Descriptor new_descriptor;
+      for (const RateLimit::DescriptorEntry& descriptorEntry : descriptor.entries_) {
+        std::string value = descriptorEntry.value_;
+
+        // TODO: do we care about case sensitivity?
+        if (strcasecmp(descriptorEntry.key_.c_str(), "remote_address") == 0 &&
+            strcasecmp(descriptorEntry.value_.c_str(), "downstream_ip") == 0) {
+          // TODO: do we want the port numbers? IPv6 support?
```

Comment:
I think we can do it step by step, if so we can reduce changes and make review easy.

Reply:
@daixiang0 so you're saying just get a happy path/simple solution in first and iterate on it in later PRs?

Reply:
/cc @envoyproxy/envoy-maintainers WDYT?

Reply:
It's fine to start with a simple approach, however once we have certain implicit behavior we can't change it, so we need to be careful to make sure whatever we do is future proof.

One thing I would like to understand is whether we should be using the substituion formatter here with a limited set of supported substitions. I think that would be better.

So basically the key is generic and set by the user, the value is optionally a substituion format string. Can we look at that and try to implement it that way?

Reply:
Here is an example of existing formatters: https://github.com/envoyproxy/envoy/blob/main/source/common/formatter/substitution_formatter.h

This will allow you to specify the descriptor value in a form %REMOTE_IP%. In this way it is a bit more flexible than using predefined key names.

Reply:
@mattklein123 @yanavlasov @rapatchi I have added some proto definitions to the API to accommodate the substitution formatters in the form of a new set of descriptors explicitly for the network rate limiter. Please let me know if this is workable or what changes need to be made. I'd like to have the API definition agreed on before I commit the time to implement this and write tests.